src/raftstore/store/snap.rs

```diff
@@ -698,135 +626,55 @@ impl Snap {
                         );
                         return Err(e);
                     }
-                    self.init_for_building(kv_snap)?;
+                    self.init_for_building()?;
                 }
             }
         }
 
-        let mut snap_key_count = 0;
         let (begin_key, end_key) = (enc_start_key(region), enc_end_key(region));
         for cf in SNAPSHOT_CFS {
             self.switch_to_cf_file(cf)?;
-            let (cf_key_count, cf_size) = if plain_file_used(cf) {
-                self.build_plain_cf_file(kv_snap, cf, &begin_key, &end_key)?
+            let cf_file = &mut self.cf_files[self.cf_index];
+            let path = cf_file.tmp_path.to_str().unwrap();
+            let cf_stat = if plain_file_used(cf_file.cf) {
+                snap_io::build_plain_cf_file(path, kv_snap, cf_file.cf, &begin_key, &end_key)?
             } else {
-                let mut key_count = 0;
-                let mut size = 0;
-                let base = self
-                    .limiter
-                    .as_ref()
-                    .map_or(0 as i64, |l| l.get_max_bytes_per_time());
-                let mut bytes: i64 = 0;
-                kv_snap.scan_cf(cf, &begin_key, &end_key, false, |key, value| {
-                    let l = key.len() + value.len();
-                    if let Some(ref limiter) = self.limiter {
-                        if bytes >= base {
-                            bytes = 0;
-                            limiter.request(base);
-                        }
-                        bytes += l as i64;
-                    }
-                    size += l;
-                    key_count += 1;
-                    box_try!(self.add_kv(key, value));
-                    Ok(true)
-                })?;
-                (key_count, size)
+                let limiter = self.limiter.as_ref().map(|c| c.as_ref());
+                snap_io::build_sst_cf_file(
+                    path, kv_snap, cf_file.cf, &begin_key, &end_key, limiter,
+                )?
             };
-            snap_key_count += cf_key_count;
+            cf_file.kv_count = cf_stat.key_count as u64;
+            if cf_file.kv_count > 0 {
```

Comment:
Please migrate the comment why we use kv_count to decide if the file is empty either than use the file size.